js_modules/dagit/packages/ui/src/components/ConfigTypeSchema.tsx

```diff
@@ -186,13 +186,123 @@ export const ConfigTypeSchema = React.memo((props: ConfigTypeSchemaProps) => {
   }
 
   return (
-    <TypeSchemaContainer>
-      <DictBlockComment content={type.description} indent="" />
-      {renderTypeRecursive(type, typeLookup, 0, props)}
-    </TypeSchemaContainer>
+    <HoveredDictEntryContextProvider>
+      <TypeSchemaContainer>
+        <DictBlockComment content={type.description} indent="" />
+        {renderTypeRecursive(type, typeLookup, 0, props)}
+      </TypeSchemaContainer>
+    </HoveredDictEntryContextProvider>
   );
 });
 
+const HoveredDictEntryContext = React.createContext<{
+  useDictEntryHover: () => {hovered: boolean; onMouseEnter: () => void; onMouseLeave: () => void};
+}>({
+  useDictEntryHover() {
+    return {hovered: false, onMouseEnter: () => {}, onMouseLeave: () => {}};
+  },
+});
+
+/**
+ * Very cheap way to make sure only 1 dict entry is hovered at a time.
+ * We simply record the unhover function for thast hovered dict entry and call it whenever
+ * a new dict entry is hovered. This is cheaper than updating every dict entry via context
+ * because we don't cause every dict entry to re-render. Only the two being hovered/unhovered.
+ */
+const HoveredDictEntryContextProvider = React.memo(({children}: {children: React.ReactNode}) => {
+  const value = React.useMemo(() => {
+    // We need to keep a stack of the entries that are hovered because they are nested.
+    // The `MouseEnter` handler only fires when we first hover the entry, but it does not
+    // fire when exiting a nested dict entry because technically we never left.
+    // To handle that case whenever we `MouseLeave` fires we restore the last element in the
+    // stack before the leaving element as hovered
+
+    let currentHoveredStack: Array<{setHovered: (hovered: boolean) => void}> = [];
+    function useDictEntryHover() {
+      const [hovered, setHovered] = React.useState(false);
```

Comment:
Hmm I feel like this stack of context -> memo -> fake hook function -> state is quite a lot to grok. (eg: what exactly re-renders when this useState is altered? I think it's the react component invoking this function, but the component must call it on EVERY render or React will throw an exception because of a "conditional" `useState`?) 

It looks structurally sound though and is pretty cool! It's elegant hiding all the mess inside this context provider. I'd worry a bit about having to re-read this to make changes / fix bugs, but it seems unlikely we'll have to do much maintenance of a behavior like this 👍 

It is also tempting to fall back to a non-React solution here though. I think this could be implemented with a few imperative methods like this, where each new mouse enter removes any other hover state, and it's just tweaking class lists after the react render... (since it'd kinda be ok for the hover state to clear when the config editor content changes, anyway.)

```
onMouseEnter={e => {
   e.currentTarget.closest('.container').querySelector('.hovering')?.classList.remove('hovering')
   e.currentTarget.classList.add('hovering')
}
onMouseLeave={e => {
   e.currentTarget.classList.remove('hovering')
}
```

Reply:
@bengotow I could use that logic where I use `setHovered(true);` but I'll still need the stack because when you exit a nested entry the parent won't get another `MouseEnter` event and then it won't look hovered even though it should.

Reply:
`context -> memo -> fake hook function -> state`
I agree its a lot but it seems necessary? 

Context -> Share state state between components

useMemo -> So we don't cause any of the components to re-render by creating a new hook every time the hover state changes

"Fake" Hook function -> Defining a hook function within a component/context is probably the easiest way to create a hook with shared state among multiple components. The alternative is to give all of the components a unique ID that they can use to make sure they all subscribe to the same state. That might actually look cleaner since now we wouldn't need to define a hook within a React component/context, however it means that we would need to use a WeakMap and ref objects that die with the component as keys to avoid keeping state around forever.

-> State -> We need the `hovered` state! 

I'm open to other approaches though this one just seemed to encapsulate everything most simply without needing to create an external store.

Reply:
Hey! Ahh I see, yeah my idea didn't address the nesting... that makes sense 👍  I think this works!

Maybe one other sorta-hacky solution -- if you put a mousemove on the sidebar container, I thikn the "event.currentTarget" would always be the container, but "event.target" would be the deepest div in the container that contained the mouse, which might be the one we want to highlight? Maaaybe 😅

Thanks for the explanation!

Reply:
@bengotow Yeah I considered that too! That would work I'd just need to check for the closest ancestor `DictEntry` and set that one as hovered. I can probably do that then not need the stack.